### ISSUE TITLE
deleting a getdns.Context() results in a defunct/zombie process

### ISSUE THREAD

**Issue by simsong** [getdnsapi/getdns-python-bindings#30]
It seems that getdns.Context() creates a subprocess that is not properly cleaned up when the context is deleted.

Example:

```
[ec2-user@ec2 ~]$ ps ux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
ec2-user 24992  0.0  0.1 113852  1968 ?        S    22:42   0:00 sshd: ec2-user@pts/0
ec2-user 24993  0.0  0.2 115400  2064 pts/0    Ss+  22:42   0:00 -bash
ec2-user 25025  0.0  0.1 113852  1968 ?        S    22:43   0:00 sshd: ec2-user@pts/1
ec2-user 25026  0.0  0.2 115400  2056 pts/1    Ss   22:43   0:00 -bash
ec2-user 25214  0.0  0.1 117168  1176 pts/1    R+   22:47   0:00 ps ux
[ec2-user@ec2 ~]$ 
```

In another window:

```
[ec2-user@ec2 ~]$ python2.7
Python 2.7.9 (default, Apr  1 2015, 18:18:03) 
[GCC 4.8.2 20140120 (Red Hat 4.8.2-16)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import getdns
>>> for i in range(100):
...     ctx = getdns.Context()
...     print(ctx.address("github.com"))
...     del ctx
...
github.com.
github.com.
github.com.
github.com.
github.com.
github.com.
github.com.
github.com.
...
github.com.
>>>
```

Back in the original window:

```
[ec2-user@ec2 ~]$ ps ux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
ec2-user 24992  0.0  0.1 113852  1968 ?        S    22:42   0:00 sshd: ec2-user@pts/0
ec2-user 24993  0.0  0.2 115400  2064 pts/0    Ss   22:42   0:00 -bash
ec2-user 25025  0.0  0.1 113852  1968 ?        S    22:43   0:00 sshd: ec2-user@pts/1
ec2-user 25026  0.0  0.2 115400  2056 pts/1    Ss   22:43   0:00 -bash
ec2-user 25215  0.2  1.5 188020 15980 pts/0    S+   22:47   0:00 python2.7
ec2-user 25216  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25217  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25218  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25219  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25220  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25221  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25222  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25223  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25224  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25225  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25226  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25227  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25228  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25229  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25230  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25231  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25232  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25233  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25234  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25235  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25236  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25237  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25238  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25239  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25240  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25241  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25242  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25243  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25244  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25245  0.0  0.0      0     0 pts/0    Z+   22:48   0:00 [python2.7] <defunct>
ec2-user 25246  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25247  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25248  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25249  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25250  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25251  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25252  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25253  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25254  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25255  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25256  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25257  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25258  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25259  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25260  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25261  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25262  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25263  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25264  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25265  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25266  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25267  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25268  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25269  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25270  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25271  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25272  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25273  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25274  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25275  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25276  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25277  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25278  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25279  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25280  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25281  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25282  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25283  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25284  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25285  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25286  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25287  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25288  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25289  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25290  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25291  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25292  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25293  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25294  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25295  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25296  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25297  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25298  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25299  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25300  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25301  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25302  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25303  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25304  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25305  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25306  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25307  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25308  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25309  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25310  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25311  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25312  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25313  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25314  0.0  0.0      0     0 pts/0    Z+   22:49   0:00 [python2.7] <defunct>
ec2-user 25315  0.0  1.3 190392 13852 pts/0    S+   22:49   0:00 python2.7
ec2-user 25316  0.0  0.1 117168  1180 pts/1    R+   22:49   0:00 ps ux
[ec2-user@ec2 ~]$ 
```

**Comment by MelindaShore** [comment]
This one surprised me a lot since I'm not doing a fork() anywhere in the Python bindings.  I put gdb on it and set a breakpoint on fork(), and the traceback gave me this

     (gdb) where
     #0  __fork () at ../nptl/sysdeps/unix/sysv/linux/pt-fork.c:25
     #1  0x00007ffff56a370b in ub_resolve_async () from /usr/local/lib/libunbound.so.2
     #2  0x00007ffff5d3a0a2 in submit_network_request (netreq=<optimized out>) at ./general.c:164
     #3  0x00007ffff5d3a3a5 in getdns_general_ns (context=<optimized out>, loop=<optimized out>, name=<optimized out>,
         request_type=<optimized out>, extensions=<optimized out>, userarg=<optimized out>, transaction_id=0x0,
         callbackfn=0x7ffff5d40fc0 <getdns_sync_cb>, internal_cb=0x0, usenamespaces=1) at ./general.c:239
     #4  0x00007ffff5d3a5c9 in priv_getdns_address_loop (context=context@entry=0xa11200,
         loop=loop@entry=0x7fffffffd9e0, name=name@entry=0x7ffff7e9ee54 "www.isoc.org",
         extensions=extensions@entry=0x0, userarg=userarg@entry=0x7fffffffd9e0,
         transaction_id=transaction_id@entry=0x0, callback=callback@entry=0x7ffff5d40fc0 <getdns_sync_cb>)
         at ./general.c:287
     #5  0x00007ffff5d41172 in getdns_address_sync (context=context@entry=0xa11200, name=0x7ffff7e9ee54 "www.isoc.org",
         extensions=extensions@entry=0x0, response=response@entry=0x7fffffffdaf8) at ./sync.c:150
     #6  0x00007ffff5f7361d in context_address (self=0x7ffff7e80140, args=<optimized out>, keywds=0x0) at context.c:1074
     #7  0x000000000049968d in PyEval_EvalFrameEx ()
     #8  0x00000000004a1634 in ?? ()
     #9  0x000000000044e87c in PyRun_InteractiveOneFlags ()
     #10 0x000000000044e998 in PyRun_InteractiveLoopFlags ()
     #11 0x000000000044ed70 in PyRun_AnyFileExFlags ()
     #12 0x000000000044f904 in Py_Main ()
     #13 0x00007ffff7817ec5 in __libc_start_main (main=0x44f9c2 <main>, argc=1, argv=0x7fffffffdfa8,
         init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdf98)
         at libc-start.c:287
     #14 0x0000000000578c4e in _start ()
     (gdb)

so basically unbound is the culprit forking a new process.  When I explicitly remove a context without exiting the process is orphaned, as well.

I'll find out what I need to do to clean up unbound processes.

**Comment by wcawijngaards** [comment]
Hi Melinda,

On 01/08/15 07:13, Melinda Shore wrote:
> This one surprised me a lot since I'm not doing a fork() anywhere in the
> Python bindings. I put gdb on it and set a breakpoint on fork(), and the
> traceback gave me this
> 
>  #1  0x00007ffff56a370b in ub_resolve_async () from /usr/local/lib/libunbound.so.2
> 
> so basically unbound is the culprit forking a new process. When I
> explicitly remove a context without exiting the process is orphaned, as
> well.
> 
> I'll find out what I need to do to clean up unbound processes.

ub_ctx_delete().

int ub_ctx_async(struct ub_ctx* ctx, int dothread) : if you set
dothread=1 it'll create a thread instead of a process.  ctx_delete calls
pthread_kill() on it.

Best regards, Wouter

**Comment by MelindaShore** [comment]
I've verified that context_dealloc() is being invoked and that it's receiving a correct context.  As a temporary workaround, unless you need a recursive resolver you can try building without unbound.  It's on its way and out that's the only thing it's being used for currently.

**Comment by simsong** [comment]
Huh? Do you have a patch with the functionality, or have you verified that the problem is not this because you are already calling the function?

We need a recursive resolver. Is unbound on the way out, to be replaced by something else, or is your patch on the way out?

**Comment by MelindaShore** [comment]
We're removing the dependency on unbound, but we're not there yet.  In the meantime, when a Context object is removed by a Python programmer or the Python interpreter the deallocate code is being called, is receiving a correct Context, and is calling the underlying libgetdns getdns_context_destroy() destructor correctly (getdns_context_destroy() does call ub_ctx_delete()).  So, I don't believe I can fix the zombie problem in the Python bindings without delving into libgetdns and libunbound and pulling the unbound context out.  I'm leery, of course, of poking around the internals of libraries that are otherwise opaque to the Python bindings, not only because of introducing dependencies on data structures that are likely to change but also because I'm likely to stomp on interactions between libgetdns and libunbound, and because I believe the deallocator is correct with respect to libgetdns.  I'm on my way out of town for several days but if this hasn't been fixed by the time I'm back I will take a look at what's going on in libgetdns.

Unbound is on its way out but I'm not sure what the timeline is for that.

**Comment by MelindaShore** [comment]
I've pushed a fix to the master branch.  Please confirm this works for you when you get a chance.

**Comment by MelindaShore** [comment]
This is going to be fixed in unbound.  I'm leaving the workaround in place in the Python bindings for the time being, to give the unbound fix time to propagate.

**Comment by simsong** [comment]
To test, I will need to move from the release to the development branch.  Let me work on it.

**Comment by MelindaShore** [comment]
This is, at bottom, a problem in libunbound, and a fix has been pushed there but until it propagates I've added a wait() to the context dealloc to clean up any exited processes that might still be around.
